# Conclusiones de la exploración inicial

In [38]:
# C00 — Cargar dataset (`pd.read_excel`)

import pandas as pd

url = "https://www.sharkattackfile.net/spreadsheets/GSAF5.xls"

df_shark_attacks = pd.read_excel(url)

df_shark_attacks.head()

Date    Year        Type            Country  \
0   10th January  2026.0  Unprovoked          Australia   
1    8th January  2026.0  Unprovoked  US Virgin Islands   
2   3rd January   2026.0  Unprovoked      New Caledonia   
3  21st December  2025.0  Unprovoked                USA   
4  12th December  2025.0  Unprovoked                USA   

                          State                     Location      Activity  \
0                           NSW                 Avalon Beach       Surfing   
1  Fredricksted Island St Croix                 Dorsch Beach    Snorkeling   
2                          Kélé  Between Bourail and Moindou  Scuba Diving   
3                    California   Lovers Point Pacific Grove      Swimming   
4      Sonoma County California                 Salmon Creek       Surfing   

            Name Sex Age  ...                     Species   \
0   Paul Stanton   M   ?  ...                      Unknown   
1  Arlene Lillis   F  56  ...                     Unknown    
2        Unknown   M   ?  ...                      Unknown   
3      Erica Fox   F  55  ...            Great White Shark   
4        Unknown   M   ?  ...  Suspected Great White Shark   

                                              Source  pdf href formula href  \
0                                     Bob Myatt GSAF  NaN          NaN  NaN   
1       Todd Smith: KevinMcMurray Trackingsharks.com  NaN          NaN  NaN   
2                         Andy Currie: Province Sud:  NaN          NaN  NaN   
3  Kevin McMurray Tracking sharks.com: Ralph Coll...  NaN          NaN  NaN   
4  Kevin McMurray Tracking sharks.com:Andrew Curr...  NaN          NaN  NaN   

  Case Number Case Number.1 original order Unnamed: 21 Unnamed: 22  
0         NaN           NaN            NaN         NaN         NaN  
1         NaN           NaN            NaN         NaN         NaN  
2         NaN           NaN            NaN         NaN         NaN  
3         NaN           NaN            NaN         NaN         NaN  
4         NaN           NaN            NaN         NaN         NaN  

[5 rows x 23 columns]

In [39]:
# C01 — Tamaño del dataset (`DataFrame.shape`)

print(df_shark_attacks.shape)

(7065, 23)


In [40]:
# C02 — Tipos + non-null por columna (`DataFrame.info`)

print(df_shark_attacks.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7065 entries, 0 to 7064
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            7065 non-null   object 
 1   Year            7063 non-null   float64
 2   Type            7047 non-null   object 
 3   Country         7015 non-null   object 
 4   State           6578 non-null   object 
 5   Location        6498 non-null   object 
 6   Activity        6480 non-null   object 
 7   Name            6846 non-null   object 
 8   Sex             6486 non-null   object 
 9   Age             4070 non-null   object 
 10  Injury          7030 non-null   object 
 11  Fatal Y/N       6504 non-null   object 
 12  Time            3538 non-null   object 
 13  Species         3934 non-null   object 
 14  Source          7045 non-null   object 
 15  pdf             6799 non-null   object 
 16  href formula    6794 non-null   object 
 17  href            6796 non-null   o

In [41]:
# C03 — Columnas numéricas (`DataFrame.select_dtypes`)

columnas_numericas = df_shark_attacks.select_dtypes(include=['number']).columns.tolist()
print (f"Las columnas numericas son : {columnas_numericas}")

Las columnas numericas son : ['Year', 'original order']


In [42]:
# C04 — Columnas categóricas (`DataFrame.select_dtypes`)

categorical_columns = df_shark_attacks.select_dtypes(include=['object']).columns.tolist()
print(f"Las columnas categoricas serian: {categorical_columns}")

Las columnas categoricas serian: ['Date', 'Type', 'Country', 'State', 'Location', 'Activity', 'Name', 'Sex', 'Age', 'Injury', 'Fatal Y/N', 'Time', 'Species ', 'Source', 'pdf', 'href formula', 'href', 'Case Number', 'Case Number.1', 'Unnamed: 21', 'Unnamed: 22']


In [43]:
# C05 — Recuentos de únicos y frecuencias + describe (función `exploration`)

def exploration(df):
    head = df.head()
    describe = df.describe().T
    info = df.info()
    describe_object = df.describe(include="object").T
    cat_cols = list(df.select_dtypes(include="object"))
    num_cols = list(df.select_dtypes(include="number"))
    for col in cat_cols:
        unique_values = df[col].nunique()
        value_counts = df[col].value_counts()
        print(f'This is {col} of unique values: {unique_values}')
        print(f'This is number of each value in {col}: {value_counts}')
    for num_col in num_cols:
        num_col_unique_values = df[num_col].nunique()
        num_col_value_counts = df[num_col].value_counts()
        print(f'This is {num_col} of unique values: {num_col_unique_values}')
        print(f'This is number of each value in {num_col}: {num_col_value_counts}')
        
    return head, describe, info, describe_object
head, describe, info, describe_object = exploration(df_shark_attacks)

describe_object

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7065 entries, 0 to 7064
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            7065 non-null   object 
 1   Year            7063 non-null   float64
 2   Type            7047 non-null   object 
 3   Country         7015 non-null   object 
 4   State           6578 non-null   object 
 5   Location        6498 non-null   object 
 6   Activity        6480 non-null   object 
 7   Name            6846 non-null   object 
 8   Sex             6486 non-null   object 
 9   Age             4070 non-null   object 
 10  Injury          7030 non-null   object 
 11  Fatal Y/N       6504 non-null   object 
 12  Time            3538 non-null   object 
 13  Species         3934 non-null   object 
 14  Source          7045 non-null   object 
 15  pdf             6799 non-null   object 
 16  href formula    6794 non-null   object 
 17  href            6796 non-null   o

count unique                                                top  \
Date           7065   6107                                               1957   
Type           7047     13                                         Unprovoked   
Country        7015    251                                                USA   
State          6578    944                                            Florida   
Location       6498   4611                   New Smyrna Beach, Volusia County   
Activity       6480   1609                                            Surfing   
Name           6846   5785                                               male   
Sex            6486     10                                                  M   
Age            4070    251                                                 16   
Injury         7030   4179                                              FATAL   
Fatal Y/N      6504     12                                                  N   
Time           3538    469                                          Afternoon   
Species        3934   1736                                        White shark   
Source         7045   5399                    K. McMurray, TrackingSharks.com   
pdf            6799   6789                          1907.10.16.R-HongKong.pdf   
href formula   6794   6784  http://sharkattackfile.net/spreadsheets/pdf_di...   
href           6796   6776  http://sharkattackfile.net/spreadsheets/pdf_di...   
Case Number    6798   6777                                         2021.07.23   
Case Number.1  6797   6775                                       2012.09.02.b   
Unnamed: 21       1      1                                       stopped here   
Unnamed: 22       2      2                                             Teramo   

               freq  
Date              9  
Type           5217  
Country        2578  
State          1192  
Location        191  
Activity       1143  
Name            679  
Sex            5665  
Age              92  
Injury          863  
Fatal Y/N      4926  
Time            215  
Species         194  
Source          131  
pdf               2  
href formula      2  
href              4  
Case Number       2  
Case Number.1     2  
Unnamed: 21       1  
Unnamed: 22       1

In [44]:
# C06 — Revisar nombres de columnas (`DataFrame.columns`)

print([columna for columna in df_shark_attacks.columns])

# Columna "Species" esta mal escrita, tiene un espacio al final

['Date', 'Year', 'Type', 'Country', 'State', 'Location', 'Activity', 'Name', 'Sex', 'Age', 'Injury', 'Fatal Y/N', 'Time', 'Species ', 'Source', 'pdf', 'href formula', 'href', 'Case Number', 'Case Number.1', 'original order', 'Unnamed: 21', 'Unnamed: 22']


In [45]:
# C07 — Porcentaje de valores nulos (`DataFrame.isnull`)

porcentage_null = df_shark_attacks.isnull().sum()/ len(df_shark_attacks)*100
print(porcentage_null)

Date               0.000000
Year               0.028309
Type               0.254777
Country            0.707714
State              6.893135
Location           8.025478
Activity           8.280255
Name               3.099788
Sex                8.195329
Age               42.392074
Injury             0.495400
Fatal Y/N          7.940552
Time              49.922151
Species           44.317056
Source             0.283086
pdf                3.765039
href formula       3.835810
href               3.807502
Case Number        3.779193
Case Number.1      3.793347
original order     3.765039
Unnamed: 21       99.985846
Unnamed: 22       99.971691
dtype: float64


In [46]:
# C08 — Duplicados (`DataFrame.duplicated`)

#Detectar duplicados
print(df_shark_attacks.duplicated().sum())

0


In [47]:
# C09 — dtype de `Year` (`Series.dtype`)

df_shark_attacks['Year'].dtype

dtype('float64')

In [48]:
# C10 — Valores de `Year` (muestra) (`Series.unique`)

print( df_shark_attacks["Year"].unique()[:100])

[2026. 2025. 2024. 2023. 2022. 2021. 2020. 2019. 2018. 2017.   nan 2016.
 2015. 2014. 2013. 2012. 2011. 2010. 2009. 2008. 2007. 2006. 2005. 2004.
 2003. 2002. 2001. 2000. 1999. 1998. 1997. 1996. 1995. 1984. 1994. 1993.
 1992. 1991. 1990. 1989. 1969. 1988. 1987. 1986. 1985. 1983. 1982. 1981.
 1980. 1979. 1978. 1977. 1976. 1975. 1974. 1973. 1972. 1971. 1970. 1968.
 1967. 1966. 1965. 1964. 1963. 1962. 1961. 1960. 1959. 1958. 1957. 1956.
 1955. 1954. 1953. 1952. 1951. 1950. 1949. 1948. 1848. 1947. 1946. 1945.
 1944. 1943. 1942. 1941. 1940. 1939. 1938. 1937. 1936. 1935. 1934. 1933.
 1932. 1931. 1930. 1929.]


In [49]:
# C11 — Nulos en `Year` (`Series.isnull().sum`)

# Detectar valores nulos en Year
print(df_shark_attacks['Year'].isnull().sum())

2


In [50]:
# C12 — Frecuencias por país (`Series.value_counts`)

print(df_shark_attacks["Country"].value_counts().head(20))

Country
USA                 2578
AUSTRALIA           1482
SOUTH AFRICA         597
NEW ZEALAND          144
PAPUA NEW GUINEA     136
BAHAMAS              136
BRAZIL               122
MEXICO               103
ITALY                 72
FIJI                  67
NEW CALEDONIA         65
PHILIPPINES           63
REUNION               60
CUBA                  49
EGYPT                 49
MOZAMBIQUE            46
SPAIN                 46
INDIA                 41
FRENCH POLYNESIA      37
JAPAN                 36
Name: count, dtype: int64


In [51]:
# C13 — Valores únicos de `Country` (`Series.unique`)

#Comprobar la forma que esta escrita los datos para saber si coinciden entre ellos aunque no esten definidos igual
#Tendremos que hacer que los que sean el mismo pais aunque esta escrito de distinta manera se agrupen
print(df_shark_attacks['Country'].unique())

['Australia' 'US Virgin Islands' 'New Caledonia' 'USA' 'French Polynesia'
 'Samoa' 'Columbia' 'Costa Rica' 'Bahamas' 'Puerto Rico' 'Spain'
 'Canary Islands' 'South Africa' 'Vanuatu' 'Jamaica' 'Israel' 'Mexico'
 'Maldives' 'Philippines' 'Turks and Caicos' 'Mozambique' 'Egypt'
 'Thailand' 'New Zealand' 'Hawaii' 'Honduras' 'Indonesia' 'Morocco'
 'Belize' 'Maldive Islands' 'Tobago' 'AUSTRALIA' 'INDIA' 'TRINIDAD'
 'BAHAMAS' 'SOUTH AFRICA' 'MEXICO' 'NEW ZEALAND' 'EGYPT' 'BELIZE'
 'PHILIPPINES' 'Coral Sea' 'SPAIN' 'PORTUGAL' 'SAMOA' 'COLOMBIA' 'ECUADOR'
 'FRENCH POLYNESIA' 'NEW CALEDONIA' 'TURKS and CaICOS' 'CUBA' 'BRAZIL'
 'SEYCHELLES' 'ARGENTINA' 'FIJI' 'MeXICO' 'ENGLAND' 'JAPAN' 'INDONESIA'
 'JAMAICA' 'MALDIVES' 'THAILAND' 'COLUMBIA' 'COSTA RICA'
 'British Overseas Territory' 'CANADA' 'JORDAN' 'ST KITTS / NEVIS'
 'ST MARTIN' 'PAPUA NEW GUINEA' 'REUNION ISLAND' 'ISRAEL' 'CHINA'
 'IRELAND' 'ITALY' 'MALAYSIA' 'LIBYA' nan 'MAURITIUS' 'SOLOMON ISLANDS'
 'ST HELENA, British overseas territory' '

In [52]:
# C14 — Ver por qué `Age` no es numérica (`Series.unique`)

# Ver por que Age no es numerico
print( df_shark_attacks["Age"].unique()[:50])

['?' '56' '55' '24' '26' '25' '61' '40' '13' '14' '50+' '54' '48' '57' '8'
 '63' '9' '39' '19' '7' '85' '69' '18' '66' '21' '37' '16' '20' '12' '42'
 '45' '30' '30+' '40+' '29' 35 58 29 24 20 55 17 12 37 36 23 40 28 69 48]


In [53]:
# C15 — `Fatal Y/N`: frecuencias + valores únicos (`Series.value_counts`, `Series.unique`)

print(f"Valores 'Y' y 'N' en: {df_shark_attacks['Fatal Y/N'].value_counts()}")

print(f"Valores unicos en Fatal Y/N: {df_shark_attacks['Fatal Y/N'].unique()}")

Valores 'Y' y 'N' en: Fatal Y/N
N          4926
Y          1486
UNKNOWN      71
 N            7
F             5
M             3
n             1
Nq            1
2017          1
Y x 2         1
N             1
y             1
Name: count, dtype: int64
Valores unicos en Fatal Y/N: ['N' 'Y' 'F' 'M' nan 'n' 'Nq' 'UNKNOWN' 2017 'Y x 2' ' N' 'N ' 'y']


In [55]:
# C16 — `Species `: nº de categorías (`Series.nunique`)

#Descripciones diferentes 
num_species = df_shark_attacks['Species '].nunique()    
print(f"El numero de especies es: {num_species}")

El numero de especies es: 1736


In [56]:
# C17 — `Species `: top valores (`Series.value_counts`)

print(df_shark_attacks['Species '].value_counts().head(30))

Species 
White shark                                           194
Shark involvement prior to death was not confirmed    105
Invalid                                               102
Shark involvement not confirmed                        92
Tiger shark                                            89
Bull shark                                             73
Shark involvement prior to death unconfirmed           68
4' shark                                               43
6' shark                                               43
1.8 m [6'] shark                                       35
Questionable incident                                  35
Questionable                                           34
1.5 m [5'] shark                                       32
5' shark                                               29
Wobbegong shark                                        27
4' to 5' shark                                         27
1.2 m [4'] shark                                       27
3' sh

In [57]:
# C18 — `Species `: ejemplos raros (`Series.sample`)

print("Ejemplos raros:")
print(df_shark_attacks['Species '].sample(10))

Ejemplos raros:
2694        Possibly a juvenile blacktip or spinner shark
3190                                                  NaN
4402                   White shark, 2.7 m [9'1"], 750-lb 
4070                                     2.4 m [8'] shark
3774                              Tiger shark, 2.1 m [7']
3910    White shark, 3.5 m [11.5'], species identity c...
3932                           White shark named “Spotty”
3585                      Shark involvement not confirmed
1693                        Oceanic whitetip shark, 2.5m 
3045                                          Tiger shark
Name: Species , dtype: object


In [59]:
# C19 — `pdf`: ¿identificador único? (`Series.nunique`)

es_unico = df_shark_attacks['pdf'].nunique() == len(df_shark_attacks)
print(f"\n¿Es la columna 'pdf' un identificador único perfecto?: {es_unico}")


¿Es la columna 'pdf' un identificador único perfecto?: False


In [60]:
# C20 — `pdf`: duplicados (`Series.duplicated().sum`)

print(f"Duplicados en pdf: {df_shark_attacks['pdf'].duplicated().sum()}")

Duplicados en pdf: 275


In [61]:
# C21 — `href`: ejemplo de valores (`DataFrame.head`)

print(df_shark_attacks[['href']].head())

  href
0  NaN
1  NaN
2  NaN
3  NaN
4  NaN


In [62]:
# C22 — `href`: nulos (`Series.isnull().sum`)

print(f"Nulos en href: {df_shark_attacks['href'].isnull().sum()}")

Nulos en href: 269


In [63]:
# C23 — `Case Number` vs `Case Number.1` (diferencias)

diferencias = df_shark_attacks[df_shark_attacks['Case Number'] != df_shark_attacks['Case Number.1']]
print(f"Cantidad de filas donde Case Number y Case Number.1 son diferentes: {len(diferencias)}")

if len(diferencias) > 0:
    print("\nMuestra de diferencias:")
    print(diferencias[['Case Number', 'Case Number.1']].head())

Cantidad de filas donde Case Number y Case Number.1 son diferentes: 317

Muestra de diferencias:
  Case Number Case Number.1
0         NaN           NaN
1         NaN           NaN
2         NaN           NaN
3         NaN           NaN
4         NaN           NaN


In [64]:
# C24 — `Unnamed: 21`: nulos (`Series.isnull().sum`)

# porcentaje de nulos
nulos_unnamed = df_shark_attacks['Unnamed: 21'].isnull().sum()
total_filas = len(df_shark_attacks)

print(f"Total filas: {total_filas}")
print(f"Nulos en Unnamed 21: {nulos_unnamed}")

Total filas: 7065
Nulos en Unnamed 21: 7064


In [66]:
# C25 — Eliminación de columnas (`DataFrame.drop`)

columnas_a_eliminar = [
    "pdf",
    "href",
    "href formula",
    "Case Number.1",
    "Unnamed: 21",
]

df_shark_attacks.drop(columns=columnas_a_eliminar, inplace=True, errors="ignore")

## Conclusiones

1. El dataset tiene `7065` filas y `23` columnas. (`(7065, 23)`)
    - Fuente (funciones): `pd.read_excel`, `DataFrame.shape` | Celdas: `C00`, `C01`

2. Hay varios tipos de datos que no son apropiados para lo que describen (y/o requieren limpieza):
    - `Date` y `Time` aparecen como `object`, pero son variables **temporales** (y además vienen en formatos mixtos).
    - `Sex` y `Fatal Y/N` aparecen como `object` y necesitan estandarización (idealmente quedarnos con categorías claras).
    - `Age` aparece como `object` pero es una variable **numérica**, pero en el dataset viene con muchos formatos distintos (rangos, “or”, “&”, etc.), así que primero hay que limpiarla y luego decidir si se convierte a número o se mantiene como categoría.
    - `Year` aparece como `float64`, pero es un año (debería ser `int` o parte de una fecha completa). Además:
        * Solo hay `2` valores nulos en `Year`.
        * Hay valores `0.0` frecuentes (probable “missing codificado”), así que conviene decidir una estrategia consistente: convertir `NaN → 0.0` por consistencia con el dataset, o convertir `0.0 → NaN` para tratarlo como missing real.
    - Porcentaje de valores nulos (aprox.):
        * `Age`: ~42%
        * `Time`: ~49%
        * `Species`: ~44%
        * `Unnamed: 21` y `Unnamed: 22`: ~99%
      Estos porcentajes hacen necesaria la valoración de si esas columnas aportan valor real a la hipótesis/análisis.
    - Fuente (funciones): `DataFrame.info`, `DataFrame.isnull`, `Series.dtype`, `Series.unique`, `Series.value_counts` | Celdas: `C02`, `C07`, `C09`, `C10`, `C11`, `C14`, `C15`

3. Valores únicos de cada columna (conclusiones de limpieza / estrategia):
    - Fuente general de recuentos: función `exploration(df)` (usa `Series.nunique()` + `Series.value_counts()`) | Celda: `C05`
    - Detección de columnas “raras” (espacios en nombre): `DataFrame.columns` | Celda: `C06`
    - Celdas específicas (solo para justificar detalles concretos):
        * `Year` (dtype/nulos/valores raros): `C09`, `C10`, `C11`
        * `Country` (top + escritura distinta): `C12`, `C13`
        * `Age` (formatos mixtos): `C14`
        * `Fatal Y/N` (recuentos + valores raros): `C15`
        * `Species ` (columna con espacio + ejemplos): `C16`, `C17`, `C18`
        * `pdf`/`href`/IDs (candidatas a eliminar): `C19`–`C23`
        * `Unnamed: 21` (nulos): `C24`

    - *Date*: `6107` valores únicos (temporal, pero está muy heterogéneo: mezcla años, fechas, rangos y `No date`).
        * 1957: 9
        * 1942: 8
        * 1958: 7
        * 1956: 6
        * No date: 6
      - Aquí lo mejor sería: normalizar siempre que sea posible y añadir una categoría difusa para los casos no normalizables (p.ej. rangos como `1845-1853`) o dejar el “raw” en una columna auxiliar.

    - *Year*: `261` valores únicos (numérica/temporal), pero hay valores `0.0` frecuentes → posible missing codificado.
        * 2015.0: 143
        * 2017.0: 141
        * 2016.0: 133
        * 0.0: 129
        * 2011.0: 128
      - Parece que la mayoría de años están bien escritos, pero conviene revisar outliers (años futuros o sin sentido) y decidir qué hacer con `0.0`.

    - *Type*: `13` valores únicos, **CATEGÓRICO** (aunque hay inconsistencias de formato: ` Provoked`, `unprovoked`, `?`, etc., que necesitan limpieza).
        * Unprovoked: 5217          -> categoría propia
        * Provoked: 641             -> categoría propia
        * Invalid: 552              -> categoría propia
        * Watercraft: 355           -> categoría propia
        * Sea Disaster: 242         -> categoría propia
        * Questionable: 26          -> categoría propia
        * Boat: 7                   -> Watercraft
        *  Provoked: 2              -> Provoked
        * unprovoked: 1             -> Unprovoked
        * ?: 1                      -> Questionable
        * Unconfirmed: 1            -> Questionable
        * Unverified: 1             -> Questionable
        * Under investigation: 1    -> Questionable
      - Estrategia: `strip()` + normalización de mayúsculas/minúsculas, y luego mapeo a un conjunto pequeño de categorías (incluyendo un “Questionable/Unknown” para ruido).

    - *Country*: `251` valores únicos, **CATEGÓRICO** (posibles duplicados semánticos por distinta escritura).
        * USA: 2578
        * AUSTRALIA: 1482
        * SOUTH AFRICA: 597
        * NEW ZEALAND: 144
        * PAPUA NEW GUINEA: 136
      - Aquí habrá que agrupar países que sean el mismo pero estén escritos distinto (normalización + mapeos).

    - *State*: `944` valores únicos, **CATEGÓRICO**, mucha variabilidad de formatos. Requiere limpieza.
        * Florida: 1192
        * New South Wales: 522
        * Queensland: 353
        * Hawaii: 345
        * California: 326
      - Estrategia: reemplazar abreviaturas a nombre completo (`AZ`, `WA`, `Cali`, etc.) y normalizar espacios/capitalización.

    - *Location*: `4611` valores únicos, **CATEGÓRICO** (texto libre / alta cardinalidad).
        * New Smyrna Beach, Volusia County: 191
        * Daytona Beach, Volusia County: 31
        * Ponce Inlet, Volusia County: 28
        * Myrtle Beach, Horry County: 22
        * Melbourne Beach, Brevard County: 20
      - Alta cardinalidad: si se usa, probablemente mediante agrupaciones (p.ej. por `Country`/`State`) o keywords.

    - *Activity*: `1609` valores únicos, **CATEGÓRICO** (alta variabilidad).
        * Surfing: 1143
        * Swimming: 1011
        * Fishing: 494
        * Spearfishing: 391
        * Wading: 178
      - Habiendo tal variabilidad, tal vez lo mejor sería crear una columna separada para guardar la información “raw” que se perderá al recategorizar. Una estrategia es recategorizar por palabras clave (`surf`, `swim`, etc.) y conservar el detalle original en otra columna.

    - *Name*: `5785` valores únicos (texto libre).
      - Es una columna con poca relevancia para la mayoría de hipótesis (no aporta magnitudes y suele ser “ruido” para conclusiones), pero se puede:
        * agrupar missing en `Unknown`
        * mantenerla solo si se necesita por contexto (p.ej. para detectar registros raros o inconsistencias), o eliminarla si no aporta a la narrativa del análisis.

    - *Sex*: `10` valores únicos, **CATEGÓRICO** (inconsistencias claras: `M`, ` M`, `m`, `F`, etc.).
        * M: 5665
        * F: 808
        * M: 4
        * F: 2
        * N: 2
        *  M: 1
        * m: 1
        * lli: 1
        * M x 2: 1
        * .: 1
      - Entre registros no nulos, hay altísima predominancia de `M`. Estrategia: limpieza a `M/F/Unknown`.

    - *Age*: `251` valores únicos, pero es `object` (formatos mixtos tipo `46 & 34`, `12 or 13`, etc.).
        * 16: 92
        * 19: 90
        * 17: 89
        * 18: 85
        * 13 or 14: 1
      - Conclusión coherente: ahora mismo **no es numérica** por cómo viene escrita; hay que limpiar formatos y luego decidir si se convierte o si se usa como categoría por rangos.

    - *Injury*: `4179` valores únicos, **CATEGÓRICO** (texto libre).
        * FATAL: 863
        * Foot bitten: 100
        * Survived: 97
        * No injury: 85
        * Leg bitten: 81
        ...
      - Se puede derivar categorías (p.ej. “fatal/no fatal”, “zona del cuerpo”, etc.), pero ya existe `Fatal Y/N` (que también requiere limpieza).

    - *Fatal Y/N*: `12` valores únicos, **CATEGÓRICO** binario esperado, pero con ruido (`UNKNOWN`, espacios, `2017`, etc.).
        * N: 4926
        * Y: 1486
        * UNKNOWN: 71
        *  N: 7
        * F: 5
      - Estrategia: normalizar y dejarlo en `Y/N/Unknown` (y revisar casos raros como números).

    - *Time*: `469` valores únicos, conceptualmente **TEMPORAL**, pero mezcla etiquetas (`Afternoon`, `Morning`) con horas (`11h00`, `15h00`, etc.).
        * Afternoon: 215
        * 11h00: 140
        * Morning: 135
        * 15h00: 127
        * 16h00: 123
        * ...
      - Necesita validación; una buena estrategia es mapear a franjas (Morning/Afternoon/Evening/Night/Unknown) y conservar el valor original por si se necesita más detalle.

    - *Species*: `1736` valores únicos, **CATEGÓRICO** (texto libre; además hay valores tipo “not confirmed” e “Invalid”).
        * White shark: 194
        * Shark involvement prior to death was not confirmed: 105
        * Invalid: 102
        * Shark involvement not confirmed: 92
        * Tiger shark: 89
      - Se puede buscar por keywords para agrupar (`white`, `tiger`, `bull`, tamaños, etc.) y mapear “not confirmed/invalid” a `Unknown`.

    - *Source*: `5399` valores únicos, **CATEGÓRICO** (texto libre / alta cardinalidad).
        * K. McMurray, TrackingSharks.com: 131
        * C. Moore, GSAF: 105
        * C. Creswell, GSAF: 101
        * S. Petersohn, GSAF: 82
        * B. Myatt, GSAF: 59
      - Normalmente aporta poco a hipótesis (suele ser “referencia”), así que es candidata a eliminar si no se usa.

    - *pdf*: `6789` valores únicos, **metadata/ID** (archivo).
      - No aporta valor directo al análisis (solo referencia al documento).

    - *href* y *href formula*: `6776` y `6784` valores únicos, **metadata/ID** (URL).
      - Indican la página web del pdf; en general no aportan al análisis (candidatas a eliminar).

    - *Case Number* y *Case Number.1*: `6777` y `6775` valores únicos, **IDENTIFICADOR**.
      - Hay duplicados puntuales (freq 2). Pueden servir para deduplicar/control de integridad, pero suelen ser candidatas a eliminar para análisis una vez validado el dataset.

    - *original order*: `6797` valores únicos (técnica/ID/orden).
      - Útil solo para preservar orden original; candidata a eliminar.

    - *Unnamed: 21* y *Unnamed: 22*: casi vacías (1–2 valores no nulos).
      - Candidatas claras a eliminar (columnas basura).

4. Duplicados:
    - No se observan duplicados (aun así, conviene validarlo con `df.duplicated().sum()` y/o deduplicar por un identificador si procede).
    - Fuente (función): `DataFrame.duplicated().sum()` | Celda: `C08`

5. Columnas útiles vs irrelevantes (orientativo; depende de la hipótesis final):
    - Columnas potencialmente útiles para hipótesis típicas:
        * Tiempo: `Year` / `Date`
        * Geografía: `Country`, `State`, `Location`
        * Demografía: `Sex`, `Age` (tras limpieza)
        * Incidente: `Fatal Y/N`, `Type`, `Activity`, `Species`
    - Columnas típicamente irrelevantes (metadata / duplicados / casi vacías):
        * `pdf`, `href`, `href formula`
        * `Case Number.1` (probable duplicado de `Case Number`)
        * `Unnamed: 21`, `Unnamed: 22`
    - Fuente: síntesis a partir de `C02`, `C05`, `C07` y checks de IDs `C19`–`C24`

6. Propuesta de columnas a eliminar (ejemplo inicial):
    - Fuente (EDA previa para decidir columnas a eliminar): `C07` + checks de IDs/columnas casi vacías `C19`–`C24`
    - Celda de eliminación: `C25`

```python
columnas_a_eliminar = [
    "pdf",
    "href",
    "href formula",
    "Case Number.1",
    "Unnamed: 21",
]

df_shark_attacks.drop(columns=columnas_a_eliminar, inplace=True, errors="ignore")
```

7. Próximos pasos (limpieza):
    - Estandarizar valores categóricos con ruido: `Type`, `Sex`, `Fatal Y/N`, `State`, `Country`.
    - Limpieza de variables temporales: `Date` y `Time` (normalización y/o categorías).
    - Definir estrategia para `Age`: limpiar formatos y convertir a numérico o a bins (rangos).
    - Revisar columnas con alta cardinalidad (`Activity`, `Location`, `Source`, `Species`) y decidir si:
        * se recategorizan por keywords, o
        * se mantienen como “raw” y se crea una versión categorizada, o
        * se eliminan si no aportan a la hipótesis.
    - Fuente: derivado de los outputs de `C02`, `C05`, `C07`, `C14`, `C15`, `C16`–`C18`